File: Session_08_MST/multi_language_implementations/kruskal_dsu.py
# Session 08: Kruskal's Algorithm with Disjoint Set Union in Python

class Edge:
    def __init__(self, src, dest, weight):
        self.src = src
        self.dest = dest
        self.weight = weight

class Graph:
    def __init__(self, V, E):
        self.V = V
        self.E = E
        self.edges = []

class Subset:
    def __init__(self, parent, rank):
        self.parent = parent
        self.rank = rank

def find(subsets, i):
    if subsets[i].parent != i:
        subsets[i].parent = find(subsets, subsets[i].parent)
    return subsets[i].parent

def union(subsets, x, y):
    xr = find(subsets, x)
    yr = find(subsets, y)
    if subsets[xr].rank < subsets[yr].rank:
        subsets[xr].parent = yr
    elif subsets[xr].rank > subsets[yr].rank:
        subsets[yr].parent = xr
    else:
        subsets[yr].parent = xr
        subsets[xr].rank += 1

def run_kruskal(graph):
    V = graph.V
    result = []
    e = 0
    i = 0
    
    # Sort edges by weight
    graph.edges.sort(key=lambda edge: edge.weight)
    
    subsets = [Subset(v, 0) for v in range(V)]
    
    while e < V - 1 and i < graph.E:
        next_edge = graph.edges[i]
        i += 1
        x = find(subsets, next_edge.src)
        y = find(subsets, next_edge.dest)
        
        if x != y:
            result.append(next_edge)
            e += 1
            union(subsets, x, y)
            
    print("Kruskal MST Edges:")
    cost = 0
    for edge in result:
        print(f"  {edge.src} - {edge.dest} == {edge.weight}")
        cost += edge.weight
    print(f"Total Cost: {cost}")

def main():
    graph = Graph(4, 5)
    graph.edges.append(Edge(0, 1, 10))
    graph.edges.append(Edge(0, 2, 6))
    graph.edges.append(Edge(0, 3, 5))
    graph.edges.append(Edge(1, 3, 15))
    graph.edges.append(Edge(2, 3, 4))
    
    run_kruskal(graph)

if __name__ == "__main__":
    main()

File: Session_08_MST/multi_language_implementations/kruskal_dsu.rs
use std::io::{self, BufRead};

#[derive(Clone, Copy)]
struct Edge {
    src: usize,
    dest: usize,
    weight: i32,
}

struct Graph {
    v: usize,
    e: usize,
    edges: Vec<Edge>,
}

struct Subset {
    parent: usize,
    rank: usize,
}

fn find(subsets: &mut [Subset], i: usize) -> usize {
    if subsets[i].parent != i {
        subsets[i].parent = find(subsets, subsets[i].parent);
    }
    subsets[i].parent
}

fn union(subsets: &mut [Subset], x: usize, y: usize) {
    let xroot = find(subsets, x);
    let yroot = find(subsets, y);

    if subsets[xroot].rank < subsets[yroot].rank {
        subsets[xroot].parent = yroot;
    } else if subsets[xroot].rank > subsets[yroot].rank {
        subsets[yroot].parent = xroot;
    } else {
        subsets[yroot].parent = xroot;
        subsets[xroot].rank += 1;
    }
}

fn run_kruskal(graph: &mut Graph) {
    let v = graph.v;
    let mut result = Vec::new();
    let mut e = 0;
    let mut i = 0;

    // Sort edges
    graph.edges.sort_by_key(|k| k.weight);

    let mut subsets = Vec::with_capacity(v);
    for idx in 0..v {
        subsets.push(Subset {
            parent: idx,
            rank: 0,
        });
    }

    println!("Kruskal's Edge Evaluation:");

    while e < v - 1 && i < graph.e {
        let next_edge = graph.edges[i];
        i += 1;

        let x = find(&mut subsets, next_edge.src);
        let y = find(&mut subsets, next_edge.dest);

        if x != y {
            result.push(next_edge);
            e += 1;
            union(&mut subsets, x, y);
            println!(
                "  Accepted edge: {} - {} (weight: {})",
                next_edge.src, next_edge.dest, next_edge.weight
            );
        } else {
            println!(
                "  Rejected edge: {} - {} (weight: {}) -> forms a cycle",
                next_edge.src, next_edge.dest, next_edge.weight
            );
        }
    }

    println!("\nFinal Minimum Spanning Tree Edges:");
    let mut minimum_cost = 0;
    for idx in 0..e {
        println!(
            "  {} - {} == {}",
            result[idx].src, result[idx].dest, result[idx].weight
        );
        minimum_cost += result[idx].weight;
    }
    println!("Total Cost of MST: {}", minimum_cost);
}

fn main() {
    let v = 4;
    let e = 5;
    let mut graph = Graph {
        v,
        e,
        edges: Vec::new(),
    };

    graph.edges.push(Edge { src: 0, dest: 1, weight: 10 });
    graph.edges.push(Edge { src: 0, dest: 2, weight: 6 });
    graph.edges.push(Edge { src: 0, dest: 3, weight: 5 });
    graph.edges.push(Edge { src: 1, dest: 3, weight: 15 });
    graph.edges.push(Edge { src: 2, dest: 3, weight: 4 });

    println!("=== Kruskal's Minimum Spanning Tree ===\n");
    run_kruskal(&mut graph);

    println!();
    println!("Press Enter to exit...");
    let mut iterator = io::stdin().lock().lines();
    let _ = iterator.next();
}
